### PR TITLE
feat(form-field): Add error state change detection

### LINKED ISSUE #1323: form-field: Improvements
- Allow with `mdc-text-field` and `mdc-select`
- Error patterns
- Sass customizations for padding and flex

### LINKED ISSUE #1549: Validators array set from controller not working
### What Angular MDC version are you using?
0.41.2
### What browser(s) is this bug affecting?
Chrome
### What OS are you using?
OS mojave
### What are the steps to reproduce the bug?
add the validators to demos/components/textfield-demo/textfield-demo.ts
```
demoForm = new FormGroup({
    username: new FormControl(
      { value: '', disabled: false }, [Validators.required, Validators.minLength(3)])
  });
```
remove them from the tag in 
demos/components/textfield-demo/textfield-demo.html
`<mdc-text-field #demoformInput formControlName="username" label="Username" outlined [helperText]="demoformHelper"></mdc-text-field>`
### What is the expected behavior?
the validation errors should display
### What is the actual behavior?
the validators messages just don't show.
### Any other information you believe would be useful?
I actually tested in the example, by changing the code in stack blitz.

### PR DESCRIPTION
**Pull request by trimox**
* Add detection of error state changes for `mdc-text-field`
* Adds `aria-invalid` attribute
* Adds `ErrorStateMatcher` class

By default, these error messages are shown when the control is invalid and either the user has interacted with (touched) the element or the parent form has been submitted. If you wish to override this behavior (e.g. to show the error as soon as the invalid control is dirty or when a parent form group is invalid), you can use the `errorStateMatcher` property of the `MdcTextField`. The property takes an instance of an `ErrorStateMatcher` object. An `ErrorStateMatcher` must implement a single method `isErrorState` which takes the `FormControl` for this `MdcTextField` as well as the parent form and returns a boolean indicating whether errors should be shown. (true indicating that they should be shown, and false indicating that they should not.)

Closes #1323 Closes #1549

### CODE REVIEW

**Comment by trimox** [comment]
Fixing unit tests..

File: packages/form-field/error-state.ts [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+  _parentForm: NgForm;
+  _defaultErrorStateMatcher: ErrorStateMatcher;
+  ngControl: NgControl;
+}
+
+/**
+ * Mixin to augment a directive with updateErrorState method.
+ * For component with `errorState` and need to update `errorState`.
+ */
+export function mixinErrorState<T extends Constructor<HasErrorState>>(base: T)
```
Comment (jpinedapetplan) [inline packages/form-field/error-state.ts]: is that a double >> ?

**Comment by trimox** [comment]
There we go..

**Comment by konraddrum** [comment]
Why is this not documented anywhere?
I spent 20 minutes figuring out what ErrorStateMatcher is [here](https://trimox.github.io/angular-mdc-web/#/angular-mdc-web/text-field-demo/examples) (in the third example).

**Comment by trimox** [comment]
Good question, and find. Opening an issue to document it.

**Pull request merged**